src/xrc/src/api.rs

```diff
@@ -476,18 +489,30 @@ async fn handle_crypto_base_fiat_quote_pair(
     num_rates_needed = num_rates_needed.saturating_add(missed_stablecoin_symbols.len());
 
     if !utils::is_caller_privileged(&caller) {
+        let current_timestamp = env.time_secs();
+        let timestamp_is_in_future = timestamp.value > current_timestamp;
         let rate_limited = is_rate_limited(num_rates_needed);
         let already_inflight = is_inflight(&request.base_asset, timestamp.value);
         let is_past_minute_not_cached =
             timestamp.r#type == NormalizedTimestampType::Past && num_rates_needed > 0;
-        let charge_cycles_option = if rate_limited || already_inflight || is_past_minute_not_cached
+        let charge_cycles_option = if rate_limited
+            || already_inflight
+            || is_past_minute_not_cached
+            || timestamp_is_in_future
         {
             ChargeOption::MinimumFee
         } else {
             ChargeOption::OutboundRatesNeeded(num_rates_needed)
         };
 
         env.charge_cycles(charge_cycles_option)?;
+        if timestamp_is_in_future {
+            return Err(errors::timestamp_is_in_future_error(
+                timestamp.value,
+                current_timestamp,
+            ));
+        }
```

Comment:
There is quite a bit of duplicate code here...

Reply:
I'm going to resolve this in a separate PR.